li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -208,7 +233,89 @@ private LiKafkaFederatedProducerImpl(LiKafkaProducerConfig configs,
-    throw new UnsupportedOperationException("Not implemented yet");
+    String topic = producerRecord.topic();
+    PerClusterProducerContext context = getOrCreateProducerContextForTopic(topic);
+    Producer<byte[], byte[]> producer = context.producer();
+    Auditor<K, V> auditor = context.auditor();
+    LongAdder numThreadsInSend = context.numThreadsInSend();
+
+    numThreadsInSend.increment();
+    boolean failed = true;
+    try {
+      if (context.closed()) {
+        throw new IllegalStateException("LiKafkaProducer has been closed.");
+      }
+
+      K key = producerRecord.key();
+      V value = producerRecord.value();
+      Object auditToken = auditor.auditToken(key, value);
+      Long timestamp = producerRecord.timestamp() == null ? System.currentTimeMillis() : producerRecord.timestamp();
+      Integer partition = producerRecord.partition();
+      Future<RecordMetadata> future = null;
+      UUID messageId = _uuidFactory.getUuid(producerRecord);
+      if (LOG.isTraceEnabled()) {
+        LOG.trace("Sending event: [{}, {}] with key {} to kafka topic {} to cluster {}",
+            messageId.toString().replaceAll("-", ""),
+            value.toString(),
+            (key != null) ? key.toString() : "[none]",
+            topic,
+            context.cluster().name());
+      }
+      byte[] serializedValue;
+      byte[] serializedKey;
+      try {
+        serializedValue = _valueSerializer.serialize(topic, value);
+        serializedKey = _keySerializer.serialize(topic, key);
+      } catch (Throwable t) {
+        // Audit the attempt. The failure will be recorded in the finally block below.
+        auditor.record(auditToken, topic, timestamp, 1L, 0L, AuditType.ATTEMPT);
+        throw t;
+      }
+      int serializedKeyLength = serializedKey == null ? 0 : serializedKey.length;
+      int serializedValueLength = serializedValue == null ? 0 : serializedValue.length;
+      int sizeInBytes = serializedKeyLength + serializedValueLength;
+      // Audit the attempt.
+      auditor.record(auditToken, topic, timestamp, 1L, (long) sizeInBytes, AuditType.ATTEMPT);
+      // We wrap the user callback for error logging and auditing purpose.
+      Callback errorLoggingCallback =
+          new ErrorLoggingCallback<>(messageId, auditToken, topic, timestamp, sizeInBytes, auditor, callback);
+      if (_largeMessageEnabled && serializedValueLength > _maxMessageSegmentSize) {
+        List<ProducerRecord<byte[], byte[]>> segmentRecords =
+            _messageSplitter.split(topic, partition, timestamp, messageId, serializedKey, serializedValue);
+        Callback largeMessageCallback = new LargeMessageCallback(segmentRecords.size(), errorLoggingCallback);
+        for (ProducerRecord<byte[], byte[]> segmentRecord : segmentRecords) {
+          future = producer.send(segmentRecord, largeMessageCallback);
+        }
+      } else if (_largeMessageSegmentWrappingRequired) {
```

Comment:
I think part of the confusion is the relationship between the configs `largeMessageEnabled` and `largeMessageSegmentWrappingRequired`. If large message support is not enabled and largeMessageSegmentWrappingRequired is set to true, it will actually wrap with large message segment. That is misleading, even though the user is not aware of these semantics.

Reply:
I agree. I think I added too much flexibility by allowing small message wrapping even when large message is not enabled and that caused confusion. 

I changed the logic a little bit so that single message wrapping is effective only when large message is enabled. I believe encryption requires large message, so this should be ok. Can you take another look? Thanks.

Reply:
Yeah. It's fine. I don't think we can improve on this without actually changing the behavior or API of the `split` method. Looks good. Thanks!